share/core.go

```diff
@@ -0,0 +1,282 @@
+	if p.g.String() != q.g.String() {
+		return nil, errorGroups
+	}
+	if p.Threshold() != q.Threshold() {
+		return nil, errorCoeffs
+	}
+	coeffs := make([]abstract.Scalar, p.Threshold())
+	for i := range coeffs {
+		coeffs[i] = p.g.Scalar().Add(p.coeffs[i], q.coeffs[i])
+	}
+	return &PriPoly{p.g, coeffs}, nil
+}
+
+// Equal checks equality of two secret sharing polynomials p and q.
+func (p *PriPoly) Equal(q *PriPoly) bool {
+	if p.g.String() != q.g.String() {
+		return false
+	}
+	b := 1
+	for i := 0; i < p.Threshold(); i++ {
+		pb := p.coeffs[i].Bytes()
+		qb := q.coeffs[i].Bytes()
+		b &= subtle.ConstantTimeCompare(pb, qb)
+	}
+	return b == 1
+}
+
+// Commit creates a public commitment polynomial for the given base point b or
+// the standard base if b == nil.
+func (p *PriPoly) Commit(b abstract.Point) *PubPoly {
+	commits := make([]abstract.Point, p.Threshold())
+	for i := range commits {
+		commits[i] = p.g.Point().Mul(b, p.coeffs[i])
+	}
+	return &PubPoly{p.g, b, commits}
+}
+
+// RecoverSecret reconstructs the shared secret p(0) from a list of private
+// shares using Lagrange interpolation.
+func RecoverSecret(g abstract.Group, shares []*PriShare, t int, n int) (abstract.Scalar, error) {
+	x := make(map[int]abstract.Scalar)
+	for i, s := range shares {
+		if s == nil || s.V == nil || s.I < 0 || n <= s.I {
+			continue
+		}
+		x[i] = g.Scalar().SetInt64(1 + int64(s.I))
+	}
+
+	if len(x) < t {
+		return nil, errors.New("not enough good private shares to reconstruct shared secret")
+	}
+
+	acc := g.Scalar().Zero()
+	num := g.Scalar()
+	den := g.Scalar()
+	tmp := g.Scalar()
+
+	for i, xi := range x {
+		num.Set(shares[i].V)
+		den.One()
+		for j, xj := range x {
+			if i == j {
+				continue
+			}
+			num.Mul(num, xj)
+			den.Mul(den, tmp.Sub(xj, xi))
+		}
+		acc.Add(acc, num.Div(num, den))
+	}
+
+	return acc, nil
+}
+
+// PubShare represents a public share.
+type PubShare struct {
+	I int            // Index of the public share
+	V abstract.Point // Value of the public share
+}
+
+// PubPoly represents a public commitment polynomial to a secret sharing polynomial.
+type PubPoly struct {
+	g       abstract.Group   // Cryptographic group
+	b       abstract.Point   // Base point, nil for standard base
+	commits []abstract.Point // Commitments to coefficients of the secret sharing polynomial
+}
+
+// NewPubPoly creates a new public commitment polynomial.
+func NewPubPoly(g abstract.Group, b abstract.Point, commits []abstract.Point) *PubPoly {
+	return &PubPoly{g, b, commits}
+}
+
+// Info returns the base point and the commitments to the polynomial coefficients.
+func (p *PubPoly) Info() (abstract.Point, []abstract.Point) {
+	return p.b, p.commits
+}
+
+// Threshold returns the secret sharing threshold.
+func (p *PubPoly) Threshold() int {
+	return len(p.commits)
+}
+
+// GetCommit returns the secret commitment p(0), i.e., the constant term of the polynomial.
+func (p *PubPoly) GetCommit() abstract.Point {
```

Comment:
Renaming to `Commit` is enough, is consistent with other methods (`Threshold` and `Info` vs `GetThreshold` and `GetInfo`) and is shorter :)